### PR TITLE
highlight: unhighlight ultra long lines of code

### PR DESCRIPTION
**Pull request by slimsag**
Someone else should take on this PR and solve the TODO section to make `go test ./internal/highlight` pass. See the added test cases and TODO comment to know how this should work. Solving this should probably take <4h total.

### CODE REVIEW

File: cmd/frontend/graphqlbackend/schema.graphql [hunk 1]
```diff
@@ -2701,7 +2701,12 @@ interface File2 {
     # The URLs to this file on external services.
     externalURLs: [ExternalLink!]!
     # Highlight the file.
-    highlight(disableTimeout: Boolean!, isLightTheme: Boolean!): HighlightedFile!
+    #
+    # If highlightLongLines is true, highlighting lines which are greater than
+    # 2000 bytes is enabled. This may produce a significant amount of HTML
+    # which some browsers (such as Chrome, but not Firefox) may have trouble
+    # rendering efficiently.
```
Comment (felixfbecker) [inline cmd/frontend/graphqlbackend/schema.graphql]: This comment would best go over the parameter itself (by wrapping parameters onto multiple lines)

File: internal/highlight/highlight.go [hunk 2]
```diff
@@ -372,3 +388,62 @@ func generatePlainTable(code string) (template.HTML, error) {
...
+// are > N bytes.
+//
+// See https://github.com/sourcegraph/sourcegraph/issues/6489
+func unhighlightLongLines(h string, n int) (string, error) {
+	doc, err := html.Parse(strings.NewReader(h))
+	if err != nil {
+		return "", err
+	}
+
+	table := doc.FirstChild.LastChild.FirstChild // html->body->table
```
Comment (felixfbecker) [inline internal/highlight/highlight.go]: Tiny suggestion, this way this is actually a well-defined CSS selector :)
```suggestion
	table := doc.FirstChild.LastChild.FirstChild // html > body > table
```

**Comment by unknwon** [comment]
Here is a preview (with lower number 50), mixed highlighted and unhighlighted lines:

image

File: internal/highlight/highlight.go (hunk 2 above)
Comment (unknwon) [reply internal/highlight/highlight.go]: Thanks, looks much better!

File: cmd/frontend/graphqlbackend/schema.graphql [hunk 3]
```diff
@@ -2701,12 +2701,15 @@ interface File2 {
...
-    #
-    # If highlightLongLines is true, highlighting lines which are greater than
-    # 2000 bytes is enabled. This may produce a significant amount of HTML
-    # which some browsers (such as Chrome, but not Firefox) may have trouble
-    # rendering efficiently.
-    highlight(disableTimeout: Boolean!, isLightTheme: Boolean!, highlightLongLines: Boolean = false): HighlightedFile!
+    highlight(
+        disableTimeout: Boolean!
+        isLightTheme: Boolean!
+        # If highlightLongLines is true, highlighting lines which are greater than
```
Comment (mrnugget) [inline cmd/frontend/graphqlbackend/schema.graphql]:
```suggestion
        # If highlightLongLines is true, lines which are longer than 2000 bytes are highlighted.
```

File: cmd/frontend/graphqlbackend/file.go [hunk 4]
```diff
@@ -93,8 +93,9 @@ func (h *highlightedFileResolver) Aborted() bool { return h.aborted }
 func (h *highlightedFileResolver) HTML() string  { return h.html }
 
 func (r *GitTreeEntryResolver) Highlight(ctx context.Context, args *struct {
-	DisableTimeout bool
-	IsLightTheme   bool
+	DisableTimeout     bool
+	IsLightTheme       bool
+	HighlightLongLines bool
```
Comment (keegancsmith) [inline cmd/frontend/graphqlbackend/file.go]: why make it possible to opt in to this? I don't see any uses of this set to true. Rather just not expose this option until we have a use case for it.

**Comment by tsenart** [comment]
@slimsag: I know you suggested we don't bother trying to address this issue in syntect_server, but architecturally, I think this is current solution is very hacky. Wouldn't it be easy to just skip lines longer than N in this loop?

https://github.com/sourcegraph/syntect_server/blob/2b5a3fb4fc9322320049563a27d9deaa815214e0/src/main.rs#L180-L184

File: cmd/frontend/graphqlbackend/file.go (hunk 4 above)
Comment (unknwon) [reply cmd/frontend/graphqlbackend/file.go]: I think @slimsag explained the rationale in https://github.com/sourcegraph/sourcegraph/issues/6489#issuecomment-584348424:

>...This is why I suggested doing it on the backend and making it an optional parameter the frontend could pass via GraphQL (in case in the future we do really want a way to say "Hey, highlight these ultra-long lines because" or maybe even having it on-by-default for Firefox which appears to handle this rendering much better). For now, it can just be off-by-default (until we get user feedback saying otherwise).

Does this solve your concern?

Comment (keegancsmith) [reply cmd/frontend/graphqlbackend/file.go]: I'd rather we add the parameter in once we have that use case, rather than speculatively. IE prefer avoid adding things to our public API.

Comment (slimsag) [reply cmd/frontend/graphqlbackend/file.go]: Another reason to have this is because there may be existing API consumers of this on e.g. Sourcegraph.com and this is a breaking change that they cannot opt-out of otherwise.

**Comment by slimsag** [comment]
@tsenart I recognize this is a 'hack', yeah. But we need to recognize that moving this into syntect_server itself isn't much better either:

- First and foremost, this is a hack to fix the fact that Chrome cannot render these lines. FIrefox doesn't have this problem, so technically speaking this should be resolved by filing an issue upstream on Chrome instead.
- If we want to argue that's not the case and it's our responsibility, then it shouldn't be in `internal/highlight` NOR should it be in `syntect_server` because then other consumers of `syntect` (not our server) also do not benefit from this improvement.
- `internal/highlight` in general, contains large amounts of code that should not exist in our Go code or `syntect_server` code, it is _all_ code that should be upstreamed to `syntect`.

In other words, our highlighting code doesn't get much love at all. If the aim here is to improve our syntax highlighting codebase overall, there are much better places we can spend our time. We already post-process the HTML we get from syntect_server in the Go code here, and this being a simple (stateless/functional) operation means we're not increasing complexity substantially either. I would rather have hacky Go code than hacky Rust code because at least then most of the team is able to refactor it.

File: cmd/frontend/graphqlbackend/file.go (hunk 4 above)
Comment (unknwon) [reply cmd/frontend/graphqlbackend/file.go]: Would `src-cli` break with the new default? Or is it using something different?

**Comment by unknwon** [comment]
This PR is pretty much self-contained and easy to revert, I'm merging it as-is.
